cranelift/docs/isle.md

```diff
@@ -0,0 +1,1306 @@
+allows ISLE rules to call into Rust in both their patterns and
+expressions, and to operate directly on types that are defined in the
+surrounding Rust code. This allows for easy and direct embedding into
+an existing compiler backend. We will cover this aspect more in the
+next section, [ISLE to Rust](#isle-to-rust).
+
+### Rules
+
+ISLE allows the user to specify rewrite rules, with a syntax similar
+in spirit to that shown above:
+
+```lisp
+    (rule
+      ;; left-hand side (pattern): if the input matches this ...
+      (A (B _ x) (C y))
+      ;; ... then rewrite to this:
+      (D x y))
+```
+
+The pattern (left-hand side) is made up of the following match
+operators:
+
+* Wildcards (`_`).
+* Integer constants (decimal/hex, positive/negative: `1`, `-1`,
+  `0x80`, `-0x80`).
+* Symbolic constants (`$Symbol`).
+* Variable captures (bare identifiers like `x`; an identifier consists
+  of alphanumeric characters and underscores, and does not start with
+  a digit).
+* Variable captures with sub-patterns: `x @ PAT`, which captures the
+  subterm in `x` as above but also matches `PAT` against the subterm.
+* "Equal-variable" constraints (`=x`): a subterm must match an
+  already-captured value.
+* conjunctions of subpatterns: `(and PAT1 PAT2 ...)` matches all of
+  the subpatterns against the term. If any subpattern does not match,
+  then this matcher fails.
+* Term deconstruction: `(term PAT1 PAT2 ...)`, where `term` is a
+  defined term (type variant or constructor) and the subpatterns are
+  applied to each argument value in turn.
+  
+The expression (right-hand side) is made up of the following
+expression operators:
+
+* Integer and symbolic constants, as above.
+* Boolean constants `#t` and `#f` (following Scheme syntax).
+* Variable uses (bare `x` identifier).
+* Term constructors (`(term EXPR1 EXPR2 ...)`, where each
+  subexpression is evaluated first and then the term is constructed).
+* `let`-expressions that bind new variables, possibly using the values
+  multiple times within the body: `(let ((var1 type1 EXPR1) (var2 ...)
+  ...) BODY ...)`. Each variable's initialization expression can refer
+  to the immediately previous variable bindings (i.e., this is like a
+  `let*` in Scheme). `let`s are lexically-scoped, meaning that bound
+  variables are available only within the body of the `let`.
+  
+When multiple rules are applicable to rewrite a particular term, ISLE
+will choose the "more specific" rule according to a particular
+heuristic: in the lowered sequence of matching steps, when one
+left-hand side completes a match while another with the same prefix
+continues with further steps, the latter (more specific) is run first.
+
+The more-specific-first heuristic is usually good enough, but when an
+undesirable choice occurs, explicit priorities can be specified.
+Rules with explicit priorities are written as `(rule PRIO lhs rhs)`
+where `PRIO` is a signed (positive or negative) integer. An applicable
```

Comment:
Does the system error if two rules with identical priorities both apply? Or fall back to the previous heuristic? What if both rules complete the match?

Reply:
There will be an error if two rules have exactly the same LHS pattern, but not otherwise; the more-specific-first heuristic breaks the tie.

I added some text about this and about how it's implementation-defined (i.e. the heuristic could change, and rules shouldn't rely on it for correctness, unlike priorities which have specifically-defined semantics).